### ISSUE TITLE
Cmake error; (newbie here)

### ISSUE THREAD

**Issue by Ub3rZ4cH** [uTox/uTox#1255]
### Overview

1. uTox version: Latest Dev (clone)
2. Operating system: Xubuntu
3. Desktop Environment/Window Manager (Linux/BSD only): XFCE

### Issue
I apologize for being a newbie I do not often build things from source, I want to install the latest uTox because the file transfer on the pre-built package is broken (0.16.1) - https://github.com/uTox/uTox/issues/1253. Someone said it was fixed in (0.17.0).

So I just attempted to build it, I created build directory and moved to it, then ran Cmake ..
At first it it was throwing an error about needing filter_audio (In dependencies.md it says filter_audio is not required but it was for me), I installed that and now I get this...
 
I get to this step and it throws this error, which I admit have no idea what it means.

> CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
DBUS_ARCH_INCLUDE_DIR
   used as include directory in directory /home/jake/Downloads/uTox-develop/src/xlib
   used as include directory in directory /home/jake/Downloads/uTox-develop/src/xlib
   used as include directory in directory /home/jake/Downloads/uTox-develop/src/xlib
   used as include directory in directory /home/jake/Downloads/uTox-develop/src/xlib
   used as include directory in directory /home/jake/Downloads/uTox-develop/src/xlib
   used as include directory in directory /home/jake/Downloads/uTox-develop/src/xlib
   used as include directory in directory /home/jake/Downloads/uTox-develop/src/xlib
   used as include directory in directory /home/jake/Downloads/uTox-develop/src/xlib
   used as include directory in directory /home/jake/Downloads/uTox-develop/src/xlib
   used as include directory in directory /home/jake/Downloads/uTox-develop/src/xlib
DBUS_INCLUDE_DIR
   used as include directory in directory /home/jake/Downloads/uTox-develop/src/xlib
   used as include directory in directory /home/jake/Downloads/uTox-develop/src/xlib
   used as include directory in directory /home/jake/Downloads/uTox-develop/src/xlib
   used as include directory in directory /home/jake/Downloads/uTox-develop/src/xlib
   used as include directory in directory /home/jake/Downloads/uTox-develop/src/xlib
   used as include directory in directory /home/jake/Downloads/uTox-develop/src/xlib
   used as include directory in directory /home/jake/Downloads/uTox-develop/src/xlib
   used as include directory in directory /home/jake/Downloads/uTox-develop/src/xlib
   used as include directory in directory /home/jake/Downloads/uTox-develop/src/xlib
   used as include directory in directory /home/jake/Downloads/uTox-develop/src/xlib

I have the latest dbus installed. 
When I run cmake again I get this...

> -- Could NOT find DBUS (missing: DBUS_INCLUDE_DIRS DBUS_LIBRARIES) 

dbus is already the newest version (1.12.2-1ubuntu1).

I only want the latest version so I can use file transfer, because like others have said, clicking the paperclip icon does nothing, and drag and drop also does nothing. When I use the screenshot function I can send a image with that work around, but not a file other than an image. However I can receive files just fine.

**Comment by eclipseo** [comment]
Have you installed the development package of dbus? It should be called libdbus-1-dev on Ubuntu.

You can see a guide here https://github.com/uTox/uTox/blob/develop/docs/BUILD.md#ubuntu on how to build for Ubuntu.

**Comment by Ub3rZ4cH** [comment]
I think that solved the Dbus error, however now I got an error saying the Third-party stuff was missing, I checked and the folders were empty (I was following the unix like guide) so I followed the ubuntu guide, and it says to download "git://github.com/uTox/uTox.git" Is this different to cloning a .zip? IDK.

Anyways the cmake worked okay, but the make threw some errors now. (sorry I am trying to understand this but I am just not knowledgeable when it comes to programming/ building from source)

So step 83% I got this...

> [ 83%] Building C object CMakeFiles/utox.dir/src/stb.c.o
In file included from /home/jake/uTox/src/stb.c:2:0:
/home/jake/uTox/src/../third-party/stb/stb_image.h: In function ‘stbi__tga_get_comp’:
/home/jake/uTox/src/../third-party/stb/stb_image.h:5242:18: warning: this statement may fall through [-Wimplicit-fallthrough=]
       case 16: if(is_grey) return STBI_grey_alpha;
                  ^
/home/jake/uTox/src/../third-party/stb/stb_image.h:5244:7: note: here
       case 15: if(is_rgb16) *is_rgb16 = 1;
       ^~~~
/home/jake/uTox/src/../third-party/stb/stb_image.h: In function ‘stbi__hdr_convert’:
/home/jake/uTox/src/../third-party/stb/stb_image.h:6466:28: warning: this statement may fall through [-Wimplicit-fallthrough=]
          case 4: output[3] = 1; /* fallthrough */
                  ~~~~~~~~~~^~~
/home/jake/uTox/src/../third-party/stb/stb_image.h:6467:10: note: here
          case 3: output[0] = output[1] = output[2] = 0;
          ^~~~
/home/jake/uTox/src/../third-party/stb/stb_image.h:6469:28: warning: this statement may fall through [-Wimplicit-fallthrough=]
          case 2: output[1] = 1; /* fallthrough */
                  ~~~~~~~~~~^~~
/home/jake/uTox/src/../third-party/stb/stb_image.h:6470:10: note: here
          case 1: output[0] = 0;
          ^~~~
In file included from /home/jake/uTox/src/stb.c:4:0:
/home/jake/uTox/src/../third-party/stb/stb_image_write.h: In function ‘stbiw__write_pixel’:
/home/jake/uTox/src/../third-party/stb/stb_image_write.h:320:13: warning: this statement may fall through [-Wimplicit-fallthrough=]
          if (!write_alpha) {
             ^
/home/jake/uTox/src/../third-party/stb/stb_image_write.h:328:7: note: here
       case 3:
       ^~~~
[ 85%] Building C object CMakeFiles/utox.dir/src/text.c.o

Then it seems to be okay and got to building the executable, but threw this lot...

> [ 98%] Building C object CMakeFiles/utox.dir/third-party/qrcodegen/c/qrcodegen.c.o
[100%] Linking C executable utox
/tmp/ccPJJjey.ltrans2.ltrans.o: In function `utox_av_local_disconnect':
<artificial>:(.text+0x442): undefined reference to `toxav_call_control'
/tmp/ccPJJjey.ltrans2.ltrans.o: In function `utox_av_local_call_control':
<artificial>:(.text+0x833): undefined reference to `toxav_call_control'
<artificial>:(.text+0x95f): undefined reference to `toxav_video_set_bit_rate'
<artificial>:(.text+0xa02): undefined reference to `toxav_video_set_bit_rate'
/tmp/ccPJJjey.ltrans2.ltrans.o: In function `utox_incoming_video_rate_change':
<artificial>:(.text+0x172d): undefined reference to `toxav_video_set_bit_rate'
/tmp/ccPJJjey.ltrans2.ltrans.o: In function `utox_incoming_audio_rate_change':
<artificial>:(.text+0x1984): undefined reference to `toxav_video_set_bit_rate'
/tmp/ccPJJjey.ltrans2.ltrans.o: In function `set_av_callbacks':
<artificial>:(.text+0x1af8): undefined reference to `toxav_callback_call'
<artificial>:(.text+0x1b10): undefined reference to `toxav_callback_call_state'
<artificial>:(.text+0x1b28): undefined reference to `toxav_callback_audio_receive_frame'
<artificial>:(.text+0x1b40): undefined reference to `toxav_callback_video_receive_frame'
<artificial>:(.text+0x1b58): undefined reference to `toxav_callback_video_bit_rate'
<artificial>:(.text+0x1b70): undefined reference to `toxav_callback_audio_bit_rate'
/tmp/ccPJJjey.ltrans2.ltrans.o: In function `utox_audio_thread':
<artificial>:(.text+0x5d29): undefined reference to `toxav_audio_send_frame'
<artificial>:(.text+0x5e83): undefined reference to `toxav_get_tox'
<artificial>:(.text+0x5f83): undefined reference to `toxav_group_send_audio'
/tmp/ccPJJjey.ltrans2.ltrans.o: In function `utox_video_thread':
<artificial>:(.text+0x8c01): undefined reference to `toxav_video_send_frame'
/tmp/ccPJJjey.ltrans3.ltrans.o: In function `utox_av_ctrl_thread':
<artificial>:(.text+0xfd68): undefined reference to `toxav_kill'
<artificial>:(.text+0x116de): undefined reference to `toxav_video_set_bit_rate'
<artificial>:(.text+0x1180a): undefined reference to `toxav_video_set_bit_rate'
<artificial>:(.text+0x11ae4): undefined reference to `toxav_iterate'
<artificial>:(.text+0x11af3): undefined reference to `toxav_iteration_interval'
<artificial>:(.text+0x11c68): undefined reference to `toxav_kill'
/tmp/ccPJJjey.ltrans5.ltrans.o: In function `utox_encrypt_data':
<artificial>:(.text+0x25264): undefined reference to `tox_pass_encrypt'
/tmp/ccPJJjey.ltrans5.ltrans.o: In function `utox_decrypt_data':
<artificial>:(.text+0x25579): undefined reference to `tox_pass_decrypt'
/tmp/ccPJJjey.ltrans5.ltrans.o: In function `load_toxcore_save':
<artificial>:(.text+0x26244): undefined reference to `tox_is_data_encrypted'
/tmp/ccPJJjey.ltrans5.ltrans.o: In function `toxcore_thread':
<artificial>:(.text+0x27996): undefined reference to `toxav_new'
/tmp/ccPJJjey.ltrans5.ltrans.o: In function `tox_thread_message':
<artificial>:(.text+0x29992): undefined reference to `toxav_call'
<artificial>:(.text+0x29add): undefined reference to `toxav_answer'
<artificial>:(.text+0x29d36): undefined reference to `toxav_answer'
<artificial>:(.text+0x29fe6): undefined reference to `toxav_add_av_groupchat'
/tmp/ccPJJjey.ltrans5.ltrans.o: In function `callback_group_invite':
<artificial>:(.text+0x2b24b): undefined reference to `toxav_join_av_groupchat'
collect2: error: ld returned 1 exit status
CMakeFiles/utox.dir/build.make:916: recipe for target 'utox' failed
make[2]: *** [utox] Error 1
CMakeFiles/Makefile2:71: recipe for target 'CMakeFiles/utox.dir/all' failed
make[1]: *** [CMakeFiles/utox.dir/all] Error 2
Makefile:151: recipe for target 'all' failed
make: *** [all] Error 2

Now btw it says Dbus found, so I think it was because I required the dev version.

**Comment by eclipseo** [comment]
For the third-party issue, this is bug #1204. Don't clone the GIT repo, you need to use the archive "uTox-0.17.0-full.tar.gz" from https://github.com/uTox/uTox/releases to get the whole 3rd party.

The message at 83% are normal (see this build I've done earlier today: https://kojipkgs.fedoraproject.org//packages/utox/0.17.0/2.fc29/data/logs/x86_64/build.log ) but the one at the end are strange. How did you build toxcore? What version are you using?

**Comment by endoffile78** [comment]
Run: 

```
git submodule init
git submodule update
```

This will download the third party repos.

**Comment by Ub3rZ4cH** [comment]
Eclipseo I built toxcore like this...

> git clone git://github.com/TokTok/c-toxcore.git
cd c-toxcore
cmake .
make
sudo checkinstall
cd ..

I dont have any problems with the third party stuff anymore, because with the link you provided for the guide, it says to download tox with **git clone --recursive git://github.com/uTox/uTox.git**

I just tried it again just to try and still no luck with make.

**Comment by Ub3rZ4cH** [comment]
Okay I apparently did not do checkinstall correctly with the toxcore, I managed to install the new toxcore.deb and also build uTox okay, but now I have an issue running it lol.

> jake@jake-HP-625:~/uTox/build$ utox
Settings:     Unable to parse utox_save.ini.
=================================================================
==10061==ERROR: AddressSanitizer: global-buffer-overflow on address 0x55c063c6a3c2 at pc 0x7f0b1d8b9f09 bp 0x7ffdc38decd0 sp 0x7ffdc38de460
WRITE of size 3 at 0x55c063c6a3c2 thread T0
    #0 0x7f0b1d8b9f08 in __interceptor_vsnprintf (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xa0f08)
    #1 0x7f0b1d8ba286 in snprintf (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xa1286)
    #2 0x55c06380c75b in config_load (/usr/local/bin/utox+0x23a75b)
    #3 0x55c063851889 in utox_init (/usr/local/bin/utox+0x27f889)
    #4 0x55c063675f27 in main (/usr/local/bin/utox+0xa3f27)
    #5 0x7f0b1d449b96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)
    #6 0x55c06364c2f9 in _start (/usr/local/bin/utox+0x7a2f9)

0x55c063c6a3c2 is located 62 bytes to the left of global variable 'edit_profile_password_data' defined in '/home/jake/uTox/src/layout/settings.c:1313:13' (0x55c063c6a400) of size 65535
0x55c063c6a3c2 is located 0 bytes to the right of global variable 'edit_video_fps_data' defined in '/home/jake/uTox/src/layout/settings.c:1312:13' (0x55c063c6a3c0) of size 2
SUMMARY: AddressSanitizer: global-buffer-overflow (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xa0f08) in __interceptor_vsnprintf
Shadow bytes around the buggy address:
  0x0ab88c785420: f9 f9 f9 f9 00 00 00 00 00 00 00 00 00 00 00 00
  0x0ab88c785430: 00 00 00 00 f9 f9 f9 f9 00 00 00 00 00 00 00 00
  0x0ab88c785440: 00 00 00 00 00 00 00 00 f9 f9 f9 f9 00 00 00 00
  0x0ab88c785450: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0ab88c785460: 00 00 00 00 00 00 00 00 00 00 00 00 f9 f9 f9 f9
=>0x0ab88c785470: 00 f9 f9 f9 f9 f9 f9 f9[02]f9 f9 f9 f9 f9 f9 f9
  0x0ab88c785480: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0ab88c785490: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0ab88c7854a0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0ab88c7854b0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0ab88c7854c0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==10061==ABORTING

**Comment by GrayHatter** [comment]
opened for reasons discussed in IRC @uTox/devs

**Comment by GrayHatter** [comment]
You were right to close it. The issue I'm talking about is somewhat unrelated to the issue you were having. But your error message is what's important.

**Comment by robinlinden** [comment]
The asan-related bug discovered in this issue has been solved in #1257. :) It is great that you opened this issue and don't hesitate to open issues in the future if there are any issues.
